scripts/evaluation.py

```diff
@@ -0,0 +1,288 @@
+import argparse
+import asyncio
+import os
+import signal
+import sys
+from asyncio import sleep
+from collections import defaultdict, deque
+from contextlib import asynccontextmanager
+from dataclasses import dataclass, field
+from pathlib import Path
+from time import perf_counter
+from typing import Any, AsyncGenerator, Deque, List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+from aiohttp import ClientSession, ServerDisconnectedError
+from cached_property import cached_property
+from cachetools import cached
+from dataclasses_json import LetterCase, dataclass_json
+
+_METRICS_WEIGHTS = pd.Series(dict(duration=-1, depth=-0.1))
+_LAMBDA = 0.6
+
+
+def cached_node_method(wrapped):
+    return cached({}, key=lambda self, *args: hash(self.sid))(wrapped)
+
+
+nitta_dataclass_params = dict(letter_case=LetterCase.CAMEL)
+
+
+@dataclass_json(**nitta_dataclass_params)
+@dataclass
+class NittaNodeDecision:
+    tag: str
+
+
+@dataclass_json(**nitta_dataclass_params)
+@dataclass
+class NittaNode:
+    score: Optional[int]
+    is_terminal: bool
+    is_finish: bool
+    parameters: Any
+    decision: NittaNodeDecision
+    duration: Optional[int]
+    sid: str
+
+    children: Optional[List["NittaNode"]] = field(default=None, repr=False)
+    parent: Optional["NittaNode"] = field(default=None, repr=False)
+
+    def __hash__(self):
+        return hash(self.sid)
+
+    @property
+    def is_leaf(self):
+        return self.is_terminal
+
+    @cached_property
+    def subtree_size(self):
+        assert self.children is not None
+        return sum(child.subtree_size for child in self.children) + 1
+
+    @cached_property
+    def depth(self) -> int:
+        return self.sid.count("-") if self.sid != "-" else 0
+
+    @cached_property
+    def subtree_leafs_metrics(self) -> Optional[Deque[Tuple[int, int]]]:
+        """:returns: deque(tuple(duration, depth)) or None if node is a failed leaf"""
+        if self.is_leaf:
+            if not self.is_finish:
+                return None
+            return deque(((self.duration, self.depth),))
+        else:
+            children_metrics = (
+                child.subtree_leafs_metrics for child in self.children if child.subtree_leafs_metrics is not None
+            )
+            return sum(children_metrics, deque())
+
+    @cached_node_method
+    def get_subtree_leafs_labels(self, metrics_distrib: np.ndarray) -> deque:
```

Comment:
Looks like get_subtree_leafs_labels and compute_label not used. Find and remove all dead code or explain why we need it.

Reply:
Those parts actually got heavily refactored in v2... Foreseeing lots of pain and conflicts...